[C-API] LLVMOrcCreateObjectLinkingLayerWithInProcessMemoryManager

macdice opened the pull request:
Allow C programs to use JITLink with trivial new C-API wrapper.  Modeled on `LLVMOrcCreateRTDyldObjectLinkingLayerWithSectionMemoryManager` except that it has to deal with failure of `jitlink::InProcessMemoryManager::Create()`.  Function name suggested by @lhames in https://github.com/llvm/llvm-project/issues/106203.

I suppose failure of underlying platform-specific things like `sysconf(_SC_PAGESIZE)` shouldn't really happen.  An alternative error reporting style might be to follow `LLVMOrcCreateDynamicLibrarySearchGeneratorForProcess` and return `LLVMErrorRef` with an output parameter for the `LLVMOrcObjectLayerRef`, but then it wouldn't be a drop-in replacement for `LLVMOrcCreateRTDyldObjectLinkingLayerWithSectionMemoryManager`.  Thoughts?

This is wanted by PostgreSQL (branch using this API: https://github.com/macdice/postgres/tree/llvm-22-proposed-c-api).  (We're also using `LLVMCreatePerfJITEventListener`, `LLVMCreatePerfJITEventListener`, `LLVMOrcRTDyldObjectLinkingLayerRegisterJITEventListener`, so it looks like we'll need to research `PerfSupportPlugin`, `DebuggerSupportPlugin`, `ObjectLinkingLayer::addPlugin()`...)

macdice committed:
[C-API] LLVMOrcCreateObjectLinkingLayerWithInProcessMemoryManager

Allow C programs to use JITLink instead of RuntimeDyld.  Modeled on
LLVMOrcCreateRTDyldObjectLinkingLayerWithSectionMemoryManager except
that it has to deal with failure to create InProcessMemoryManager.  In
that case it returns nullptr.

lhames commented:
> Does this all make sense? Thanks!

Yep -- that all makes sense.

> I suppose failure of underlying platform-specific things like sysconf(_SC_PAGESIZE) shouldn't really happen. An alternative error reporting style might be to follow LLVMOrcCreateDynamicLibrarySearchGeneratorForProcess and return LLVMErrorRef with an output parameter for the LLVMOrcObjectLayerRef, but then it wouldn't be a drop-in replacement for LLVMOrcCreateRTDyldObjectLinkingLayerWithSectionMemoryManager. Thoughts?

I think it's good to have a direct error reporting path where possible, so
```c
LLVMErrorRef LLVMOrcCreateObjectLinkingLayerWithInProcessMemoryManager(
    LLVMOrcInProcessMemoryManagerRef &MemMgr, LLVMOrcExecutionSessionRef ES);
```
would be preferred.

In the function body `mm` should be capitalised per LLVM conventions, but otherwise LGTM.

macdice committed:
amend! [C-API] LLVMOrcCreateObjectLinkingLayerWithInProcessMemoryManager

[C-API] LLVMOrcCreateObjectLinkingLayerWithInProcessMemoryManager (#169862)

Allow C programs to use JITLink instead of RuntimeDyld.  Modeled on
LLVMOrcCreateRTDyldObjectLinkingLayerWithSectionMemoryManager except
that it has to deal with failure to create InProcessMemoryManager.

macdice committed:
fixup! amend! [C-API] LLVMOrcCreateObjectLinkingLayerWithInProcessMemoryManager

macdice commented:
Thanks!  Sorry for the bad commit message, I didn't realise that my patch's commit message would be replaced with the PR message.  At least I'll know that next time...

lhames commented:
No worries -- I didn't think to reword it before merging either. ;)
